### PR TITLE
feat: add autocomplete

### LINKED ISSUE #503: feat: add user-based auto-complete endpoint
- [x] research auto completion libraries and choose the best one (importance to the fact the we can add our add vocabulary perhaps use tags) 
1. use preconfiguration
2. use index data
3. start from scratch use search data
- [x] implement executor // to all the other executors in the flow (add tests)
- [x] `/search` endpoint to update user queries with any search
- [x] `/get_suggestion` to get auto-completion suggestions for given query (use library for this)
- [x] add `get_suggestion` to bff for apps with input modality text
- [x] add end to end test
- [x] update playground 
- [x] add documentation about the feature

### PR DESCRIPTION
**Pull request by azayz**
closes #503

### CODE REVIEW

File: tests/now_executors/NOWAutoCompleteExecutor/test_unit.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+def test_search_update_profanity():
+    executor = NOWAutoCompleteExecutor()
+
+    da = DocumentArray(
+        [
+            Document(text='background'),
+            Document(text='background'),
+            Document(text='shit'),
+            Document(text='loading'),
+            Document(text='fuck'),
```
Comment (makram93) [inline tests/now_executors/NOWAutoCompleteExecutor/test_unit.py:81]: Maybe you can code word this too, `f*ck` instead, just like `l3sbi4n`

**Comment by makram93** [comment]
Are we also planning to implement autocorrection of the search term as well?

**Review by joschkabraun — CHANGES_REQUESTED** [review]
If I understand your PR correctly, then the auto suggestion only works on completing a word given its beginning letters, is that correct? Can you extend that functionality to also include suggesting next words which are often follow the current one?

File: now/apps/music_to_music/app.py [hunk 2]
```diff
@@ -106,6 +106,9 @@ def setup(
             ] = f"jinahub+docker://{pre_trained_head_map[user_input.data]}"
 
             self.set_flow_yaml(demo_data=True)
+            super().setup(
```
[the remark this answers is not in the document]
Comment (azayz) [reply now/apps/music_to_music/app.py:109]: it won't work if the flow_yaml isn't set

File: now/apps/text_to_text_and_image/flow.yml [hunk 3]
```diff
@@ -56,6 +56,11 @@ executors:
     select_model: clip-vision
   env:
     JINA_LOG_LEVEL: DEBUG
+  - name: autocomplete_executor
```
[the remark this answers is not in the document]
Comment (azayz) [reply now/apps/text_to_text_and_image/flow.yml:59]: good catch

File: now/apps/text_to_video/flow-video-clip.yml [hunk 4]
```diff
@@ -55,5 +55,10 @@ executors:
       resources:
         memory: ${{ ENV.INDEXER_MEM }}
         cpu: 1.0
+    env:
+      JINA_LOG_LEVEL: DEBUG
+  - name: autocomplete_executor
```
[the remark this answers is not in the document]
Comment (azayz) [reply now/apps/text_to_video/flow-video-clip.yml:60]: good catch

File: now/apps/text_to_video/flow-video-clip-high-performance.yml [hunk 5]
```diff
@@ -62,5 +62,10 @@ executors:
       resources:
         memory: 4G
         cpu: 1.0
+    env:
+      JINA_LOG_LEVEL: DEBUG
+  - name: autocomplete_executor
```
[the remark this answers is not in the document]
Comment (azayz) [reply now/apps/text_to_video/flow-video-clip-high-performance.yml:67]: good catch

File: now/apps/text_to_video/ft-flow-video-clip.yml [hunk 6]
```diff
@@ -75,5 +75,10 @@ executors:
       resources:
         memory: ${{ ENV.INDEXER_MEM }}
         cpu: 1.0
+    env:
+      JINA_LOG_LEVEL: DEBUG
+  - name: autocomplete_executor
```
[the remark this answers is not in the document]
Comment (azayz) [reply now/apps/text_to_video/ft-flow-video-clip.yml:80]: good catch

File: now/apps/text_to_video/ft-flow-video-clip-high-performance.yml [hunk 7]
```diff
@@ -78,5 +78,10 @@ executors:
       resources:
         memory: 8G
         cpu: 2.0
+    env:
+      JINA_LOG_LEVEL: DEBUG
+  - name: autocomplete_executor
```
[the remark this answers is not in the document]
Comment (azayz) [reply now/apps/text_to_video/ft-flow-video-clip-high-performance.yml:83]: good catch

File: now/apps/music_to_music/app.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (joschkabraun) [reply now/apps/music_to_music/app.py:109]: flow yaml will be set in `common_setup`

File: tests/now_executors/NOWAutoCompleteExecutor/test_unit.py (hunk 1 above)
Comment (azayz) [reply tests/now_executors/NOWAutoCompleteExecutor/test_unit.py:81]: added

**Comment by girishc13** [comment]
I have two `suggestions`:

- From the api perspective limit the size of characters or words using a threshold. This will prevent issues due to crawling attacks or misuse of the api. It also doesn't make sense to provide suggestions for very long word or sentence.
- You can also use [this](https://github.com/minimaxir/big-list-of-naughty-strings) resource to improve the integration or e2e testing for the api.

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -335,7 +389,7 @@ jobs:
 
   executor-test:
     runs-on: ubuntu-latest
-    needs: [ update-preprocessor-if-required, update-qdrant-executor-if-required, update-bff-playground-if-required, update-elastic-executor-if-required]
```
Comment (violenil) [inline .github/workflows/ci.yml:338]: How come these executor tests (for qdrant and elastic) are no longer run?

File: deployment/bff/app/v1/routers/txt2video.py [hunk 9]
```diff
@@ -77,3 +77,21 @@ def search(data: NowTextSearchRequestModel):
...
+    Return text suggestions for the rest of the query text
+    """
+    query_doc, _ = process_query(text=data.text, uri=data.uri, conditions=data.filters)
+
+    docs = jina_client_post(
+        data=data,
+        inputs=query_doc,
+        endpoint='/suggestion',
+    )
+    return docs.to_dict()
```
Comment (violenil) [inline deployment/bff/app/v1/routers/txt2video.py:97]: Does it make sense to have a 'base router' file that holds functions that all routers can use? To reduce the amount of duplicate code here.

File: now/executor/autocomplete/executor.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+from better_profanity import profanity
+from docarray import DocumentArray
+from fast_autocomplete import AutoComplete
+
+from now.executor.abstract.auth.auth import NOWAuthExecutor as Executor
+from now.executor.abstract.auth.auth import SecurityLevel, secure_request
+
+
+class NOWAutoCompleteExecutor(Executor):
+    def __init__(self, search_traversal_paths: str = '@r', words=None, *args, **kwargs):
```
Comment (violenil) [inline now/executor/autocomplete/executor.py:12]: are `words` usually of type `Dict`?

File: now/executor/autocomplete/executor.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    ):
+        traversal_paths = parameters.get('traversal_paths', self.search_traversal_paths)
+        flat_docs = docs[traversal_paths]
+
+        for doc in flat_docs:
+            if doc.text:
+                search_words = doc.text.split(' ')
+                # in case query is composed of two words
+                for word in search_words:
+                    if not NOWAutoCompleteExecutor.contains_profanity(word):
```
Comment (violenil) [inline now/executor/autocomplete/executor.py:40]: I think we should still allow users to search using some profane terms if they wish to, but we should not cache them or use them for our autocompletion. Are we disregarding profane words altogether? Just for my own understanding.

**Review by violenil — COMMENTED** [review]
Left some questions, mostly for my own understanding :) Looks great!

File: deployment/bff/app/v1/routers/txt2video.py (hunk 9 above)
Comment (azayz) [reply deployment/bff/app/v1/routers/txt2video.py:97]: you mean a function where e put the `process_query` and `jina client post` functions

File: deployment/bff/app/v1/routers/text2text.py [hunk 12]
```diff
@@ -61,3 +61,22 @@ def search(data: NowTextSearchRequestModel):
     )
 
     return docs[0].matches.to_dict()
+
+
+@router.post(
+    "/suggestion",
```
[the remark this answers is not in the document]
Comment (azayz) [reply deployment/bff/app/v1/routers/text2text.py:67]: how about `/autocomplete`

File: deployment/bff/app/v1/routers/txt2video.py (hunk 9 above)
Comment (makram93) [reply deployment/bff/app/v1/routers/txt2video.py:97]: This endpoint is limited to only `text` input modality, hence we can not add to base.

File: deployment/playground/src/search.py [hunk 13]
```diff
@@ -87,6 +86,25 @@ def search(attribute_name, attribute_value, jwt, top_k=None, filter_dict=None):
     return call_flow(URL_HOST, data, attribute_name, domain)
 
 
+def get_suggestion(text, jwt):
```
Comment (makram93) [inline deployment/playground/src/search.py:89]: Maybe you can omit this function and reuse the search function adding an additional param to it `endpoints` which specifies the host end?
```suggestion
def get_suggestion(text, jwt, endpoint='search'):
.
.
.
URL_HOST = = f"{domain}/api/v1/{params.input_modality}-to-{params.output_modality}/{endpoints}"
```

File: now/executor/autocomplete/executor.py [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+    @secure_request(on='/search', level=SecurityLevel.USER)
+    def search_update(
+        self, docs: Optional[DocumentArray] = None, parameters: dict = {}, **kwargs
+    ):
+        traversal_paths = parameters.get('traversal_paths', self.search_traversal_paths)
+        flat_docs = docs[traversal_paths]
+
+        for doc in flat_docs:
+            if doc.text and not NOWAutoCompleteExecutor.contains_profanity(doc.text):
```
Comment (makram93) [inline now/executor/autocomplete/executor.py:43]: `contains_profinity` returns True if it contains some profanity right? Then shouldn't this check be? :
```suggestion
            if doc.text and NOWAutoCompleteExecutor.contains_profanity(doc.text):
```

**Review by makram93 — COMMENTED** [review]
Overall looks good. Nice work. 
Are we going ahead with the playground changes for now and incorporate the custom components as and when it is ready?

File: now/executor/autocomplete/executor.py (hunk 14 above)
Comment (makram93) [reply now/executor/autocomplete/executor.py:43]: Ah okay, never mind. I got what this is doing. So you add the searched words if it is not profane one.

File: deployment/playground/src/search.py (hunk 13 above)
Comment (azayz) [reply deployment/playground/src/search.py:94]: applied

File: now/executor/autocomplete/executor.py (hunk 11 above)
Comment (violenil) [reply now/executor/autocomplete/executor.py]: Great!

File: docs/user-guides/production.md [hunk 15]
```diff
@@ -33,6 +33,11 @@ Then you can provide your `AWS access key ID` and `AWS secret key`, you can prov
 Your data is automatically loaded from the S3 bucket. It will be temporary stored on EC2 machines while processing the data
 but it will not be stored.
 
+## Auto-complete suggestions
+
+For aplications with Text as a modality input, we provide `autocomplete` feature. This feature learns slowly from your users's search requests
```
Comment (violenil) [inline docs/user-guides/production.md:38]:
```suggestion
For applications with Text as input modality, we provide the `autocomplete` feature. This feature learns slowly from the user's search requests
```

**Comment by azayz** [comment]
Screenshot from 2022-10-20 13-40-15

File: docs/user-guides/production.md (hunk 15 above)
Comment (azayz) [reply docs/user-guides/production.md:38]: Thanks, applied

File: deployment/bff/app/v1/routers/txt2video.py (hunk 9 above)
Comment (joschkabraun) [reply deployment/bff/app/v1/routers/txt2video.py:97]: Once we refactor bff into the apps, we should do that by checking in the baseclass if `input_modality == 'text'`

File: docs/user-guides/production.md [hunk 16]
```diff
@@ -33,6 +33,14 @@ Then you can provide your `AWS access key ID` and `AWS secret key`, you can prov
 Your data is automatically loaded from the S3 bucket. It will be temporary stored on EC2 machines while processing the data
 but it will not be stored.
 
+## Auto-complete suggestions
+
+For applications with Text as input modality, we provide the `autocomplete` feature. This feature learns slowly from the user's search requests
```
Comment (joschkabraun) [inline docs/user-guides/production.md:38]: text is lower case

File: now/executor/autocomplete/executor.py [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+from docarray import DocumentArray
+from fast_autocomplete import AutoComplete
+
+from now.executor.abstract.auth.auth import NOWAuthExecutor as Executor
+from now.executor.abstract.auth.auth import SecurityLevel, secure_request
+
+
+class NOWAutoCompleteExecutor(Executor):
+    def __init__(self, search_traversal_paths: str = '@r', words=None, *args, **kwargs):
+        self.words = words if words else {}
```
Comment (joschkabraun) [inline now/executor/autocomplete/executor.py:13]: We should persist `words` to the workspace similar to how the authentication executor does it with any secrets. The issue is we lose the state if the executor restarts.

File: now/executor/autocomplete/executor.py [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+                # prevent users from misusing API
+                if len(doc.text) < self.char_threshhold:
+                    # in case query is composed of two words
+                    for word in search_words:
+                        self.update_words(word)
+                    # add bigram and tri gram suggestions
+                    if len(search_words) == 2 or len(search_words) == 3:
+                        self.update_words(doc.text)
+
+        self.auto_complete = AutoComplete(words=self.words)
```
Comment (joschkabraun) [inline now/executor/autocomplete/executor.py:54]: What's the introduced latency of that object creation for a very large corpus? Do you have any numbers on how long that takes for words to contain millions of words?

File: now/executor/autocomplete/requirements.txt [hunk 19]
```diff
@@ -0,0 +1,5 @@
+fast-autocomplete==0.9.0
+Levenshtein==0.20.5
```
Comment (joschkabraun) [inline now/executor/autocomplete/requirements.txt:2]: do we use levenhstein requirements?

File: now/executor/autocomplete/executor.py (hunk 14 above)
Comment (joschkabraun) [inline now/executor/autocomplete/executor.py:43]: Given that `contains_profanity` is only one line of code, I don't think we need an extra function for that.

File: requirements-test.txt [hunk 20]
```diff
@@ -7,4 +7,8 @@ pytest-reraise
 pytest-custom_exit_code
 mock
 black==22.3.0
-pytest-split==0.8.0
\ No newline at end of file
+pytest-split==0.8.0
+fast-autocomplete==0.9.0
+Levenshtein==0.20.5
+python-Levenshtein==0.20.5
```
Comment (joschkabraun) [inline requirements-test.txt:13]: do we use levenhstein requirements?

**Review by joschkabraun — APPROVED** [review]
LGTM! Some minor comments.

File: now/executor/autocomplete/requirements.txt (hunk 19 above)
Comment (azayz) [reply now/executor/autocomplete/requirements.txt:2]: yes it's needed for levenshtein python

File: requirements-test.txt (hunk 20 above)
Comment (azayz) [reply requirements-test.txt:13]: yes needed for fast autocomplete, I tried removing and I got errors

File: .github/workflows/ci.yml [hunk 21]
```diff
@@ -335,7 +389,7 @@ jobs:
...
+    needs: [ update-preprocessor-if-required, update-bff-playground-if-required,  update-autocomplete-executor-if-required, update-qdrant-executor-if-required, update-elastic-executor-if-required ] 
     steps:
       - uses: actions/checkout@v2
       - name: Set up Python 3.7
```
Comment (florian-hoenicke) [inline .github/workflows/ci.yml:395]: Please add test run for Autocomplete

File: now/executor/abstract/base_indexer/base_indexer.py [hunk 22]
```diff
@@ -66,6 +66,10 @@ def get_tags_and_values(self, **kwargs):
         count_dict = defaultdict(lambda: defaultdict(int))
         for tags in self.doc_id_tags.values():
             for key, value in tags.items():
+                if not (
```
[the remark this answers is not in the document]
Comment (florian-hoenicke) [reply now/executor/abstract/base_indexer/base_indexer.py:69]: Maybe it is not needed anymore? Or can be fixed at another level in another pr

**Review by florian-hoenicke — CHANGES_REQUESTED** [review]
The fixes for tags can be done in a separated pr to have it handled in the right way

**Comment by azayz** [comment]
Screenshot from 2022-10-26 09-15-30
Screenshot from 2022-10-26 09-14-57

**Comment by makram93** [comment]
@azayz Can we move the `search` button to align with the search bar? You can use columns to keep them side-by-side. This way we can move the suggestion up to be close to the search bar.

**Review by violenil — APPROVED** [review]
LGTM! Good job 🚀

**Review by florian-hoenicke — APPROVED** [review]
Well done. This is a big feature!

**Pull request merged**